mm2src/lp_price.rs

```
@@ -0,0 +1,340 @@
+                Some(rate_infos)
+            },
+            None => None,
+        }
+    }
+}
+
+pub async fn process_price_request(price_url: &str) -> Result<TickerInfosRegistry, MmError<PriceServiceRequestError>> {
+    debug!("Fetching price from: {}", price_url);
+    let (status, headers, body) = slurp_url(price_url).await?;
+    let (status_code, body, _) = (status, std::str::from_utf8(&body)?.trim().into(), headers);
+    if status_code != StatusCode::OK {
+        return MmError::err(PriceServiceRequestError::HttpProcessError(body));
+    }
+    let model: HashMap<String, TickerInfos> = serde_json::from_str(&body)?;
+    Ok(TickerInfosRegistry(model))
+}
+
+pub async fn fetch_price_tickers(price_url: &str) -> Result<TickerInfosRegistry, MmError<PriceServiceRequestError>> {
+    let model = process_price_request(price_url).await?;
+    debug!("price registry size: {}", model.0.len());
+    Ok(model)
+}
+
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct CEXRates {
+    pub base: Option<BigDecimal>,
+    pub rel: Option<BigDecimal>,
+}
+// Fetcher funtion to test fetching latest prices from a endoint.
+async fn try_price_fetcher_endpoint(
+    endpoint: &str,
+    base: &str,
+    rel: &str,
+) -> Result<CEXRates, MmError<PriceServiceRequestError>> {
+    match process_price_request(endpoint).await {
+        Ok(response) => match response.get_cex_rates(base.to_string(), rel.to_string()) {
+            Some(fiat_price) => {
+                let (base_usd_price, rel_usd_price) = fiat_price.get_rate_price();
+                Ok(CEXRates {
+                    base: Some(base_usd_price),
+                    rel: Some(rel_usd_price),
+                })
+            },
+            None => {
+                return MmError::err(PriceServiceRequestError::Internal(
+                    "Error encoutered while trying to get cex rates.".to_string(),
+                ))
+            },
```

If there is need to return an error on `None`, you can use `or_mm_err`:
https://github.com/KomodoPlatform/atomicDEX-API/blob/b434af8025829d4f570971f12f465c854784f2ec/mm2src/common/mm_error/or_mm_error.rs#L13-L22